File: pbark/pbark/parser/expression/condition_parser.py
from __future__ import annotations

from dataclasses import dataclass

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import keywords as Keywords
from pbark.parser.expression.comparison_op import ComparisonOp
from pbark.parser.parse_expression import (
    Comparison,
    Field,
    HasExact,
    HasTrait,
    Logical,
    LogicalOp,
    Not,
    NumberLiteral,
    ParseExpression,
)


@dataclass
class _CountAndTopic:
    amount: float
    topic: str


class ConditionParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_condition(self, start: int, end: int, line: int) -> ParseExpression:
        if self._word_at(start, Keywords.NEITHER_WORDS):
            return self._parse_neither_condition(start, end, line)
        return self._parse_and_joined_tests(start, end, line)

    def _parse_neither_condition(self, start: int, end: int, line: int) -> ParseExpression:
        nor_at = self._find_keyword_index(start + 1, end, Keywords.NOR_WORDS)
        if nor_at < 0:
            raise BarkError(
                line,
                "This command without nor really confused the dogs... They are doing random stuff now!",
            )
        first_dog = self._find_breed_name(start + 1, nor_at)
        second_dog = self._find_breed_name(nor_at + 1, end)
        count = self._read_count_and_topic(nor_at + 1, end)
        if first_dog is None or second_dog is None or count is None:
            raise BarkError(
                line, "Either one of the dogs or the toy is missing! Did you look around?"
            )
        first_has = HasExact(first_dog, count.topic, count.amount)
        second_has = HasExact(second_dog, count.topic, count.amount)
        return Logical(first_has, LogicalOp.NOR, second_has)

    def _parse_and_joined_tests(self, start: int, end: int, line: int) -> ParseExpression:
        combined: ParseExpression | None = None
        test_start = start
        while test_start < end:
            and_at = self._find_keyword_index(test_start, end, Keywords.CONDITION_AND_WORDS)
            test_end = and_at if and_at >= 0 else end
            one_test = self._parse_one_test(test_start, test_end, line)
            if combined is None:
                combined = one_test
            else:
                combined = Logical(combined, LogicalOp.AND, one_test)
            if and_at < 0:
                return combined
            test_start = and_at + 1
        return combined  # type: ignore[return-value]

    def _parse_one_test(self, start: int, end: int, line: int) -> ParseExpression:
        test = self._match_more_than_count(start, end)
        if test is not None:
            return test
        test = self._match_sniffs_less_than(start, end)
        if test is not None:
            return test
        test = self._match_trait_check(start, end)
        if test is not None:
            return test
        test = self._match_exact_count(start, end)
        if test is not None:
            return test
        raise BarkError(
            line,
            "The dogs tilted their heads at you as they have no clue what you want them to do!",
        )

    def _match_more_than_count(self, start: int, end: int) -> ParseExpression | None:
        for offset in range(start, end - 1):
            if not self._is_more_than_phrase(offset):
                continue
            who = self._find_dog_name(start, offset)
            limit: float | None = None
            topic: str | None = None
            for scan in range(offset + 2, end):
                token = self._at(scan)
                if limit is None and token.is_(TokenType.NUMBER):
                    limit = float(self._parser.normalise(token.value))
                elif limit is not None and topic is None and token.is_(TokenType.IDENTIFIER):
                    resolved = Keywords.explicit_attribute_topic(
                        self._parser.normalise(token.value)
                    )
                    if resolved is not None:
                        topic = resolved
            if who is None or limit is None or topic is None:
                return None
            compare = Comparison(Field(who, topic), ComparisonOp.GREATER_THAN, NumberLiteral(limit))
            if self._is_negated_before_count(start, end):
                return Not(compare)
            return compare
        return None

    def _match_sniffs_less_than(self, start: int, end: int) -> ParseExpression | None:
        sniffs_at = self._find_keyword_index(start, end, Keywords.SNIFFS_WORDS)
        if sniffs_at < 0:
            return None
        who = self._find_dog_name(start, sniffs_at)
        scan = sniffs_at + 1
        if scan >= end or not self._word_at(scan, Keywords.COMPARISON_LESS_WORDS):
            return None
        scan += 1
        if scan >= end or not self._at(scan).is_(TokenType.IDENTIFIER):
            return None
        topic = Keywords.explicit_attribute_topic(self._parser.normalise(self._at(scan).value))
        if topic is None:
            return None
        scan += 1
        if scan >= end or not self._word_at(scan, Keywords.THAN_WORDS):
            return None
        scan += 1
        if scan >= end or not self._at(scan).is_(TokenType.NUMBER):
            return None
        if who is None:
            return None
        limit = float(self._parser.normalise(self._at(scan).value))
        compare = Comparison(Field(who, topic), ComparisonOp.LESS_THAN, NumberLiteral(limit))
        if self._is_negated_before_count(start, end):
            return Not(compare)
        return compare

    def _match_trait_check(self, start: int, end: int) -> ParseExpression | None:
        breed = self._find_breed_name(start, end)
        trait_at = self._find_trait_word(start, end)
        if breed is None or trait_at < 0 or self._has_number_in_range(start, end):
            return None
        trait = self._parser.normalise(self._at(trait_at).value)
        check = HasTrait(breed, trait)
        if self._is_negated_before(trait_at, start):
            return Not(check)
        return check

    def _match_exact_count(self, start: int, end: int) -> ParseExpression | None:
        who = self._find_dog_name(start, end)
        count = self._read_count_and_topic(start, end)
        if who is None or count is None:
            return None
        exact = HasExact(who, count.topic, count.amount)
        if self._is_negated_before_count(start, end):
            return Not(exact)
        return exact

    def _read_count_and_topic(self, start: int, end: int) -> _CountAndTopic | None:
        amount: float | None = None
        topic: str | None = None
        for offset in range(start, end):
            token = self._at(offset)
            if amount is None and token.is_(TokenType.NUMBER):
                amount = float(self._parser.normalise(token.value))
                continue
            if amount is not None and topic is None and token.is_(TokenType.IDENTIFIER):
                resolved = Keywords.explicit_attribute_topic(
                    self._parser.normalise(token.value)
                )
                if Keywords.is_attribute_keyword(resolved):
                    topic = resolved
        if amount is None or topic is None:
            return None
        return _CountAndTopic(amount, topic)

    def _is_negated_before_count(self, start: int, end: int) -> bool:
        for offset in range(start, end):
            if self._at(offset).is_(TokenType.NUMBER):
                return False
            if self._word_at(offset, Keywords.NOT_KEYWORDS):
                return True
        return False

    def _is_negated_before(self, before: int, start: int) -> bool:
        for offset in range(start, before):
            if self._word_at(offset, Keywords.NOT_KEYWORDS):
                return True
        return False

    def _find_keyword_index(self, from_: int, end: int, words: list[str]) -> int:
        for offset in range(from_, end):
            if self._word_at(offset, words):
                return offset
        return -1

    def _find_dog_name(self, start: int, end: int) -> str | None:
        for offset in range(start, end):
            token = self._at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                continue
            word = self._parser.normalise(token.value)
            if Keywords.is_dog_subject_reference(word, token.value):
                return word
        return None

    def _find_breed_name(self, start: int, end: int) -> str | None:
        for offset in range(start, end):
            token = self._at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                continue
            word = self._parser.normalise(token.value)
            if ConfigLoader.is_breed(word):
                return word
        return None

    def _find_trait_word(self, start: int, end: int) -> int:
        for offset in range(start, end):
            token = self._at(offset)
            if token.is_(TokenType.IDENTIFIER) and Keywords.is_trait_keyword(
                self._parser.normalise(token.value)
            ):
                return offset
        return -1

    def _has_number_in_range(self, start: int, end: int) -> bool:
        for offset in range(start, end):
            if self._at(offset).is_(TokenType.NUMBER):
                return True
        return False

    def _is_more_than_phrase(self, offset: int) -> bool:
        return self._word_at(offset, Keywords.COMPARISON_GREATER_WORDS) and self._word_at(
            offset + 1, Keywords.THAN_WORDS
        )

    def _word_at(self, offset: int, words: list[str]) -> bool:
        token = self._at(offset)
        return token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words

    def _at(self, offset: int):
        return self._parser.peek_at(offset)

File: src/main/java/dev/klomptech/jbark/parser/expression/ConditionParser.java
package dev.klomptech.jbark.parser.expression;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;
import java.util.List;

// Example: when she has 2 toys then
// Example: when she has 3 toys and my labrador is loud then
// Example: when my labrador does not have 9 toys then
// Example: when neither my labrador nor my beagle has 9 toys then
public class ConditionParser {

  private final Parser parser;

  // amount + topic from "9 toys"
  private record CountAndTopic(double amount, String topic) {}

  public ConditionParser(final Parser parser) {
    this.parser = parser;
  }

  // Words between when and then
  public ParseExpression parseCondition(final int start, final int end, final int line)
      throws BarkError {
    if (wordAt(start, Keywords.NEITHER_WORDS)) {
      return parseNeitherCondition(start, end, line);
    }
    return parseAndJoinedTests(start, end, line);
  }

  // Example: "neither my labrador nor my beagle has 9 toys"
  private ParseExpression parseNeitherCondition(final int start, final int end, final int line)
      throws BarkError {
    int norAt = findKeywordIndex(start + 1, end, Keywords.NOR_WORDS);
    if (norAt < 0) {
      throw new BarkError(
          line,
          "This command without nor really confused the dogs... They are doing random stuff now!");
    }
    String firstDog = findBreedName(start + 1, norAt);
    String secondDog = findBreedName(norAt + 1, end);
    CountAndTopic count = readCountAndTopic(norAt + 1, end);
    if (firstDog == null || secondDog == null || count == null) {
      throw new BarkError(
          line, "Either one of the dogs or the toy is missing! Did you look around?");
    }
    ParseExpression firstHas =
        new ParseExpression.HasExact(firstDog, count.topic(), count.amount());
    ParseExpression secondHas =
        new ParseExpression.HasExact(secondDog, count.topic(), count.amount());
    return new ParseExpression.Logical(firstHas, ParseExpression.LogicalOp.NOR, secondHas);
  }

  // One test, or several joined by 'and'. Each part ends at 'and' or at 'then'
  private ParseExpression parseAndJoinedTests(final int start, final int end, final int line)
      throws BarkError {
    ParseExpression combined = null;
    int testStart = start;
    while (testStart < end) {
      int andAt = findKeywordIndex(testStart, end, Keywords.CONDITION_AND_WORDS);
      int testEnd = andAt >= 0 ? andAt : end;
      // Read one test between when and then (or one half before 'and')
      ParseExpression oneTest = parseOneTest(testStart, testEnd, line);
      if (combined == null) {
        combined = oneTest;
      } else {
        combined = new ParseExpression.Logical(combined, ParseExpression.LogicalOp.AND, oneTest);
      }
      // No 'and' on the line, one test only
      if (andAt < 0) {
        return combined;
      }
      testStart = andAt + 1;
    }
    return combined;
  }

  // Pick which phrase this part is. Order matters, 'more than' before 'has'
  private ParseExpression parseOneTest(final int start, final int end, final int line)
      throws BarkError {
    ParseExpression test = matchMoreThanCount(start, end);
    if (test != null) {
      return test;
    }
    // Example: "she sniffs less toys than 8"
    test = matchSniffsLessThan(start, end);
    if (test != null) {
      return test;
    }
    // Example: "my labrador is loud"
    test = matchTraitCheck(start, end);
    if (test != null) {
      return test;
    }
    // Example: "she has 2 toys"
    test = matchExactCount(start, end);
    if (test != null) {
      return test;
    }
    throw new BarkError(
        line, "The dogs tilted their heads at you as they have no clue what you want them to do!");
  }

  // "she has more than 2 toys"
  private ParseExpression matchMoreThanCount(final int start, final int end) {
    for (int offset = start; offset < end - 1; offset++) {
      // Example: "she has more than 2 toys"
      if (!isMoreThanPhrase(offset)) {
        continue;
      }
      // Pull out who, limit, and topic
      String who = findDogName(start, offset);
      Double limit = null;
      String topic = null;
      for (int scan = offset + 2; scan < end; scan++) {
        Token token = at(scan);
        if (limit == null && token.is(TokenType.NUMBER)) {
          limit = Double.valueOf(parser.normalise(token.value()));
        } else if (limit != null && topic == null && token.is(TokenType.IDENTIFIER)) {
          String resolved = Keywords.explicitAttributeTopic(parser.normalise(token.value()));
          if (resolved != null) {
            topic = resolved;
          }
        }
      }
      if (who == null || limit == null || topic == null) {
        return null;
      }
      // Create the comparison expression
      ParseExpression compare =
          new ParseExpression.Comparison(
              new ParseExpression.Field(who, topic),
              ComparisonOp.GREATER_THAN,
              new ParseExpression.NumberLiteral(limit));
      if (isNegatedBeforeCount(start, end)) {
        return new ParseExpression.Not(compare);
      }
      return compare;
    }
    return null;
  }

  // "she sniffs less toys than 8", find 'sniffs' first, then read the words after it
  private ParseExpression matchSniffsLessThan(final int start, final int end) {
    int sniffsAt = findKeywordIndex(start, end, Keywords.SNIFFS_WORDS);
    if (sniffsAt < 0) {
      return null;
    }
    String who = findDogName(start, sniffsAt);
    int scan = sniffsAt + 1;
    // After 'sniffs', we expect 'less'
    if (scan >= end || !wordAt(scan, Keywords.COMPARISON_LESS_WORDS)) {
      return null;
    }
    scan++;
    // After 'less', we expect a specific attribute topic
    if (scan >= end || !at(scan).is(TokenType.IDENTIFIER)) {
      return null;
    }
    String topic = Keywords.explicitAttributeTopic(parser.normalise(at(scan).value()));
    if (topic == null) {
      return null;
    }
    scan++;
    // After the topic, we expect 'than'
    if (scan >= end || !wordAt(scan, Keywords.THAN_WORDS)) {
      return null;
    }
    scan++;
    // After 'than', we expect a number
    if (scan >= end || !at(scan).is(TokenType.NUMBER)) {
      return null;
    }
    if (who == null) {
      return null;
    }
    double limit = Double.parseDouble(parser.normalise(at(scan).value()));
    ParseExpression compare =
        new ParseExpression.Comparison(
            new ParseExpression.Field(who, topic),
            ComparisonOp.LESS_THAN,
            new ParseExpression.NumberLiteral(limit));
    // Example: "she does not sniff less toys than 8"
    if (isNegatedBeforeCount(start, end)) {
      return new ParseExpression.Not(compare);
    }
    return compare;
  }

  // "my labrador is loud" or "my labrador is not greedy"
  private ParseExpression matchTraitCheck(final int start, final int end) {
    String breed = findBreedName(start, end);
    int traitAt = findTraitWord(start, end);
    if (breed == null || traitAt < 0 || hasNumberInRange(start, end)) {
      return null;
    }
    String trait = parser.normalise(at(traitAt).value());
    ParseExpression check = new ParseExpression.HasTrait(breed, trait);
    if (isNegatedBefore(traitAt, start)) {
      return new ParseExpression.Not(check);
    }
    return check;
  }

  // "she has 2 toys" or "she does not have 9 toys"
  private ParseExpression matchExactCount(final int start, final int end) {
    String who = findDogName(start, end);
    // Example: "she has 2 toys"
    CountAndTopic count = readCountAndTopic(start, end);
    if (who == null || count == null) {
      return null;
    }
    ParseExpression exact = new ParseExpression.HasExact(who, count.topic(), count.amount());
    // Example: "she does not have 9 toys"
    if (isNegatedBeforeCount(start, end)) {
      return new ParseExpression.Not(exact);
    }
    return exact;
  }

  // Pulls "9 toys" out of a word range
  private CountAndTopic readCountAndTopic(final int start, final int end) {
    Double amount = null;
    String topic = null;
    for (int offset = start; offset < end; offset++) {
      Token token = at(offset);
      // Example: "she has 2 toys"
      if (amount == null && token.is(TokenType.NUMBER)) {
        amount = Double.valueOf(parser.normalise(token.value()));
        continue;
      }
      if (amount != null && topic == null && token.is(TokenType.IDENTIFIER)) {
        String resolved = Keywords.explicitAttributeTopic(parser.normalise(token.value()));
        if (Keywords.isAttributeKeyword(resolved)) {
          topic = resolved;
        }
      }
    }
    if (amount == null || topic == null) {
      return null;
    }
    return new CountAndTopic(amount, topic);
  }

  // "does not have 9", 'not' appears before the number
  private boolean isNegatedBeforeCount(final int start, final int end) {
    for (int offset = start; offset < end; offset++) {
      // Example: "she does not have 9 toys"
      if (at(offset).is(TokenType.NUMBER)) {
        return false;
      }
      if (wordAt(offset, Keywords.NOT_KEYWORDS)) {
        return true;
      }
    }
    return false;
  }

  // "is not greedy", 'not' appears before the trait
  private boolean isNegatedBefore(final int before, final int start) {
    for (int offset = start; offset < before; offset++) {
      if (wordAt(offset, Keywords.NOT_KEYWORDS)) {
        return true;
      }
    }
    return false;
  }

  private int findKeywordIndex(final int from, final int end, final List<String> words) {
    for (int offset = from; offset < end; offset++) {
      if (wordAt(offset, words)) {
        return offset;
      }
    }
    return -1;
  }

  private String findDogName(final int start, final int end) {
    for (int offset = start; offset < end; offset++) {
      Token token = at(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.isDogSubjectReference(word, token.value())) {
        return word;
      }
    }
    return null;
  }

  private String findBreedName(final int start, final int end) {
    for (int offset = start; offset < end; offset++) {
      Token token = at(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (ConfigLoader.isBreed(word)) {
        return word;
      }
    }
    return null;
  }

  private int findTraitWord(final int start, final int end) {
    for (int offset = start; offset < end; offset++) {
      Token token = at(offset);
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.isTraitKeyword(parser.normalise(token.value()))) {
        return offset;
      }
    }
    return -1;
  }

  private boolean hasNumberInRange(final int start, final int end) {
    for (int offset = start; offset < end; offset++) {
      if (at(offset).is(TokenType.NUMBER)) {
        return true;
      }
    }
    return false;
  }

  private boolean isMoreThanPhrase(final int offset) {
    return wordAt(offset, Keywords.COMPARISON_GREATER_WORDS)
        && wordAt(offset + 1, Keywords.THAN_WORDS);
  }

  private boolean wordAt(final int offset, final List<String> words) {
    Token token = at(offset);
    return token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()));
  }

  private Token at(final int offset) {
    return parser.peekAt(offset);
  }
}
